File: emerald-streaming-windows/publish/services/obsStreamProbeService.js
const fs = require("node:fs");
const os = require("node:os");
const path = require("node:path");
const { spawn } = require("node:child_process");

async function probeObsStream(request) {
  if (!request.inputUrl || !String(request.inputUrl).trim()) {
    throw new Error("OBS recording URL is required before stream stats can be probed.");
  }

  const ffprobePath = normalizeFfprobePath(request.ffmpegPath);
  const output = await runFfprobe(ffprobePath, [
    "-v", "error",
    "-select_streams", "v:0",
    "-show_entries", "stream=codec_name,width,height,avg_frame_rate,r_frame_rate,bit_rate:format=bit_rate",
    "-of", "json",
    String(request.inputUrl).trim(),
  ]);

  return parseResult(output);
}

function runFfprobe(ffprobePath, args) {
  return new Promise((resolve, reject) => {
    const process = spawn(ffprobePath, args, {
      windowsHide: true,
      stdio: ["ignore", "pipe", "pipe"],
    });

    let output = "";
    let errorOutput = "";
    const timeout = setTimeout(() => {
      process.kill("SIGKILL");
      reject(new Error("ffprobe timed out while reading the OBS stream. Make sure OBS is streaming and the URL is reachable."));
    }, 8000);

    process.stdout.on("data", (chunk) => {
      output += chunk.toString();
    });

    process.stderr.on("data", (chunk) => {
      errorOutput += chunk.toString();
    });

    process.on("error", (error) => {
      clearTimeout(timeout);
      reject(new Error(`Unable to start ffprobe at '${ffprobePath}'. ${error.message}`));
    });

    process.on("exit", (code) => {
      clearTimeout(timeout);
      if (code !== 0) {
        reject(new Error(errorOutput.trim() || "ffprobe could not read the stream."));
        return;
      }

      resolve(output);
    });
  });
}

function parseResult(json) {
  const root = JSON.parse(json);
  const stream = Array.isArray(root.streams) && root.streams.length > 0 ? root.streams[0] : {};
  const codec = stream.codec_name || null;
  const width = toInt(stream.width);
  const height = toInt(stream.height);
  const fps = parseFrameRate(stream.avg_frame_rate) || parseFrameRate(stream.r_frame_rate);
  const bitrate = toInt(stream.bit_rate) || toInt(root.format?.bit_rate);

  return {
    codec,
    width,
    height,
    fps,
    bitrate,
    bitrateText: formatBitrate(bitrate),
    resolutionText: formatResolution(width, height),
  };
}

function normalizeFfprobePath(configuredPath) {
  if (!configuredPath || !String(configuredPath).trim()) {
    return "ffprobe";
  }

  const trimmedPath = String(configuredPath).trim().replace(/^"|"$/g, "");

  if (fs.existsSync(trimmedPath) && fs.statSync(trimmedPath).isDirectory()) {
    return path.join(trimmedPath, os.platform() === "win32" ? "ffprobe.exe" : "ffprobe");
  }

  const fileName = path.basename(trimmedPath);
  if (fileName.toLowerCase().startsWith("ffmpeg")) {
    const directory = path.dirname(trimmedPath);
    return directory === "." ? (os.platform() === "win32" ? "ffprobe.exe" : "ffprobe") : path.join(directory, os.platform() === "win32" ? "ffprobe.exe" : "ffprobe");
  }

  return trimmedPath;
}

function toInt(value) {
  if (value === null || value === undefined || value === "") {
    return null;
  }

  const parsed = Number.parseInt(value, 10);
  return Number.isNaN(parsed) ? null : parsed;
}

function parseFrameRate(frameRate) {
  if (!frameRate || frameRate === "0/0") {
    return null;
  }

  const parts = String(frameRate).split("/");
  if (parts.length === 2) {
    const numerator = Number.parseFloat(parts[0]);
    const denominator = Number.parseFloat(parts[1]);

    if (!Number.isNaN(numerator) && !Number.isNaN(denominator) && denominator > 0) {
      return numerator / denominator;
    }
  }

  const value = Number.parseFloat(frameRate);
  return Number.isNaN(value) ? null : value;
}

function formatBitrate(bitrate) {
  if (!bitrate) {
    return "Not reported";
  }

  return bitrate >= 1000000
    ? `${trimNumber(bitrate / 1000000)} Mbps`
    : `${trimNumber(bitrate / 1000)} Kbps`;
}

function formatResolution(width, height) {
  if (!width || !height) {
    return "Not reported";
  }

  let quality;
  if (height >= 2160) quality = "2160p";
  else if (height >= 1440) quality = "1440p";
  else if (height >= 1080) quality = "1080p";
  else if (height >= 720) quality = "720p";
  else if (height >= 480) quality = "480p";
  else quality = `${height}p`;

  return `${width} x ${height} (${quality})`;
}

function trimNumber(value) {
  return Number(value.toFixed(2)).toString();
}

module.exports = {
  probeObsStream,
};

File: Emerald.Streaming/ObsStreamProbeService.cs
using System.Diagnostics;
using System.Globalization;
using System.Text.Json;

namespace Emerald.Streaming;

public static class ObsStreamProbeService
{
    public static async Task<ObsStreamProbeResult> ProbeAsync(ObsStreamProbeRequest request, CancellationToken cancellationToken)
    {
        if (string.IsNullOrWhiteSpace(request.InputUrl))
        {
            throw new InvalidOperationException("OBS recording URL is required before stream stats can be probed.");
        }

        var ffprobePath = NormalizeFfprobePath(request.FfmpegPath);
        using var timeout = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
        timeout.CancelAfter(TimeSpan.FromSeconds(8));

        var startInfo = new ProcessStartInfo
        {
            FileName = ffprobePath,
            UseShellExecute = false,
            RedirectStandardOutput = true,
            RedirectStandardError = true,
            CreateNoWindow = true
        };

        AddArguments(startInfo, request.InputUrl.Trim());

        using var process = Process.Start(startInfo)
            ?? throw new InvalidOperationException($"Unable to start ffprobe at '{ffprobePath}'.");

        var outputTask = process.StandardOutput.ReadToEndAsync(timeout.Token);
        var errorTask = process.StandardError.ReadToEndAsync(timeout.Token);

        try
        {
            await process.WaitForExitAsync(timeout.Token);
        }
        catch (OperationCanceledException)
        {
            TryKill(process);
            throw new InvalidOperationException("ffprobe timed out while reading the OBS stream. Make sure OBS is streaming and the URL is reachable.");
        }

        var output = await outputTask;
        var error = await errorTask;

        if (process.ExitCode != 0)
        {
            throw new InvalidOperationException(string.IsNullOrWhiteSpace(error) ? "ffprobe could not read the stream." : error.Trim());
        }

        return ParseResult(output);
    }

    private static void AddArguments(ProcessStartInfo startInfo, string inputUrl)
    {
        startInfo.ArgumentList.Add("-v");
        startInfo.ArgumentList.Add("error");
        startInfo.ArgumentList.Add("-select_streams");
        startInfo.ArgumentList.Add("v:0");
        startInfo.ArgumentList.Add("-show_entries");
        startInfo.ArgumentList.Add("stream=codec_name,width,height,avg_frame_rate,r_frame_rate,bit_rate:format=bit_rate");
        startInfo.ArgumentList.Add("-of");
        startInfo.ArgumentList.Add("json");
        startInfo.ArgumentList.Add(inputUrl);
    }

    private static ObsStreamProbeResult ParseResult(string json)
    {
        using var document = JsonDocument.Parse(json);
        var root = document.RootElement;
        var stream = root.TryGetProperty("streams", out var streams) && streams.GetArrayLength() > 0
            ? streams[0]
            : default;

        var codec = GetString(stream, "codec_name");
        var width = GetInt(stream, "width");
        var height = GetInt(stream, "height");
        var fps = ParseFrameRate(GetString(stream, "avg_frame_rate")) ?? ParseFrameRate(GetString(stream, "r_frame_rate"));
        var streamBitrate = GetLong(stream, "bit_rate");
        var formatBitrate = root.TryGetProperty("format", out var format) ? GetLong(format, "bit_rate") : null;
        var bitrate = streamBitrate ?? formatBitrate;

        return new ObsStreamProbeResult(
            codec,
            width,
            height,
            fps,
            bitrate,
            FormatBitrate(bitrate),
            FormatResolution(width, height));
    }

    private static string NormalizeFfprobePath(string? configuredPath)
    {
        if (string.IsNullOrWhiteSpace(configuredPath))
        {
            return "ffprobe";
        }

        var trimmedPath = configuredPath.Trim().Trim('"');

        if (Directory.Exists(trimmedPath))
        {
            return Path.Combine(trimmedPath, OperatingSystem.IsWindows() ? "ffprobe.exe" : "ffprobe");
        }

        var fileName = Path.GetFileName(trimmedPath);
        if (fileName.StartsWith("ffmpeg", StringComparison.OrdinalIgnoreCase))
        {
            var directory = Path.GetDirectoryName(trimmedPath);
            return string.IsNullOrWhiteSpace(directory)
                ? (OperatingSystem.IsWindows() ? "ffprobe.exe" : "ffprobe")
                : Path.Combine(directory, OperatingSystem.IsWindows() ? "ffprobe.exe" : "ffprobe");
        }

        return trimmedPath;
    }

    private static string? GetString(JsonElement element, string propertyName)
    {
        return element.ValueKind == JsonValueKind.Object
            && element.TryGetProperty(propertyName, out var property)
            && property.ValueKind == JsonValueKind.String
                ? property.GetString()
                : null;
    }

    private static int? GetInt(JsonElement element, string propertyName)
    {
        return element.ValueKind == JsonValueKind.Object
            && element.TryGetProperty(propertyName, out var property)
            && property.TryGetInt32(out var value)
                ? value
                : null;
    }

    private static long? GetLong(JsonElement element, string propertyName)
    {
        if (element.ValueKind != JsonValueKind.Object || !element.TryGetProperty(propertyName, out var property))
        {
            return null;
        }

        if (property.ValueKind == JsonValueKind.Number && property.TryGetInt64(out var numberValue))
        {
            return numberValue;
        }

        return property.ValueKind == JsonValueKind.String
            && long.TryParse(property.GetString(), NumberStyles.Integer, CultureInfo.InvariantCulture, out var stringValue)
                ? stringValue
                : null;
    }

    private static double? ParseFrameRate(string? frameRate)
    {
        if (string.IsNullOrWhiteSpace(frameRate) || frameRate == "0/0")
        {
            return null;
        }

        var parts = frameRate.Split('/');
        if (parts.Length == 2
            && double.TryParse(parts[0], NumberStyles.Float, CultureInfo.InvariantCulture, out var numerator)
            && double.TryParse(parts[1], NumberStyles.Float, CultureInfo.InvariantCulture, out var denominator)
            && denominator > 0)
        {
            return numerator / denominator;
        }

        return double.TryParse(frameRate, NumberStyles.Float, CultureInfo.InvariantCulture, out var value) ? value : null;
    }

    private static string FormatBitrate(long? bitrate)
    {
        if (bitrate is null)
        {
            return "Not reported";
        }

        return bitrate >= 1_000_000
            ? $"{bitrate.Value / 1_000_000d:0.##} Mbps"
            : $"{bitrate.Value / 1_000d:0.##} Kbps";
    }

    private static string FormatResolution(int? width, int? height)
    {
        if (width is null || height is null)
        {
            return "Not reported";
        }

        var quality = height switch
        {
            >= 2160 => "2160p",
            >= 1440 => "1440p",
            >= 1080 => "1080p",
            >= 720 => "720p",
            >= 480 => "480p",
            _ => $"{height}p"
        };

        return $"{width} x {height} ({quality})";
    }

    private static void TryKill(Process process)
    {
        try
        {
            if (!process.HasExited)
            {
                process.Kill(entireProcessTree: true);
            }
        }
        catch
        {
            // Best-effort cleanup after timeout.
        }
    }
}

public sealed record ObsStreamProbeRequest(string InputUrl, string? FfmpegPath);

public sealed record ObsStreamProbeResult(
    string? Codec,
    int? Width,
    int? Height,
    double? Fps,
    long? Bitrate,
    string BitrateText,
    string ResolutionText);
